File: src/main/java/io/scriptor/impl/MMU.java
package io.scriptor.impl;

import com.carrotsearch.hppc.ObjectArrayList;
import com.carrotsearch.hppc.ObjectObjectHashMap;
import com.carrotsearch.hppc.ObjectObjectMap;
import io.scriptor.machine.Hart;
import io.scriptor.util.Log;
import org.jetbrains.annotations.NotNull;

import static io.scriptor.isa.CSR.*;

public final class MMU {

    private record Key(long vpn, long asid) {
    }

    private static final class Entry {

        long pteaddr;
        long pte;
        long vpn;
        long asid;
        long pgbase;
        long pgsize;

        private Entry(
                final long pteaddr,
                final long pte,
                final long vpn,
                final long asid,
                final long pgbase,
                final long pgsize
        ) {
            this.pteaddr = pteaddr;
            this.pte = pte;
            this.vpn = vpn;
            this.asid = asid;
            this.pgbase = pgbase;
            this.pgsize = pgsize;
        }

        public boolean contains(final long vpn) {
            final var pgcount = pgsize >>> PAGE_SHIFT;
            return this.vpn <= vpn && vpn < this.vpn + pgcount;
        }
    }

    public enum Access {
        FETCH,
        READ,
        WRITE,
    }

    private static final long FETCH_PAGE_FAULT = 0x0CL;
    private static final long READ_PAGE_FAULT = 0x0DL;
    private static final long WRITE_PAGE_FAULT = 0x0FL;

    private static final int PAGE_SHIFT = 12;

    private final Hart hart;

    private final ObjectObjectMap<Key, Entry> tlb = new ObjectObjectHashMap<>();

    public MMU(final @NotNull Hart hart) {
        this.hart = hart;
    }

    public void flush(final long vaddr, final long asid) {
        if (vaddr == 0L && asid == 0L) {
            tlb.clear();
            return;
        }

        final var vpn = vaddr >>> PAGE_SHIFT;

        final var remove = new ObjectArrayList<Key>();
        for (final var e : tlb) {
            final var matches = (asid == 0L || asid == e.key.asid) && (vaddr == 0L || e.value.contains(vpn));
            if (matches) {
                remove.add(e.key);
            }
        }

        for (final var key : remove) {
            tlb.remove(key.value);
        }
    }

    public long translate(
            final int priv,
            final long vaddr,
            final @NotNull Access access,
            final boolean unsafe
    ) {
        if (priv == CSR_M) {
            return vaddr;
        }

        final var reg  = hart.csrFile().getd(satp, priv);
        final var mode = getMode(reg);
        final var asid = getASID(reg);
        final var ppn  = getPPN(reg);

        return switch (mode) {
            case 0 -> vaddr;
            case 8 -> sv39(priv, vaddr, access, asid, ppn, unsafe);
            case 9 -> sv48(priv, vaddr, access, asid, ppn, unsafe);
            case 10 -> sv57(priv, vaddr, access, asid, ppn, unsafe);
            default -> {
                pageFault(priv, vaddr, access, unsafe, "unsupported mode %d".formatted(mode));
                yield ~0L;
            }
        };
    }

    private long touch(
            final int priv,
            final long vaddr,
            final @NotNull Access access,
            final boolean unsafe,
            final long pteaddr,
            long pte
    ) {
        if (inaccessible(priv, access, pte)) {
            pageFault(priv, vaddr, access, unsafe, "inaccessible");
        }

        if (unprivileged(priv, pte)) {
            pageFault(priv, vaddr, access, unsafe, "unprivileged");
        }

        if (!pteA(pte) || (access == Access.WRITE && !pteD(pte))) {
            pte |= (1L << 6);
            if (access == Access.WRITE) {
                pte |= (1L << 7);
            }
            hart.machine().pWrite(pteaddr, 8, pte, unsafe);
        }

        return pte;
    }

    private long walk(
            final int levels,
            final int priv,
            final long vaddr,
            final @NotNull Access access,
            final long asid,
            final long root,
            final boolean unsafe
    ) {
        {
            final var vpn = vaddr >>> PAGE_SHIFT;

            var entry = tlb.get(new Key(vpn, asid));
            if (entry == null) {
                entry = tlb.get(new Key(vpn, 0L));
            }
            if (entry != null && entry.contains(vpn)) {
                entry.pte = touch(priv, vaddr, access, unsafe, entry.pteaddr, entry.pte);

                final var pgoffset = vaddr & (entry.pgsize - 1L);
                return entry.pgbase | pgoffset;
            }
        }

        var a = root << PAGE_SHIFT;

        Log.info("walk(levels=%d, priv=%x, vaddr=%x, access=%s, asid=%x, root=%x)",
                 levels,
                 priv,
                 vaddr,
                 access,
                 asid,
                 root);

        for (int i = levels - 1; i >= 0; --i) {
            final var vpn = vpn(vaddr, i);

            Log.info("  i=%d, a=%x, vpn=%x", i, a, vpn);

            final var pteaddr = a + vpn * 8L;

            var pte = hart.machine().pRead(pteaddr, 8, unsafe);

            Log.info("   => pteaddr=%x, pte=%016x, V=%b, R=%b, W=%b, X=%b, U=%b, G=%b, A=%b, D=%b",
                     pteaddr,
                     pte,
                     pteV(pte),
                     pteR(pte),
                     pteW(pte),
                     pteX(pte),
                     pteU(pte),
                     pteG(pte),
                     pteA(pte),
                     pteD(pte));

            if (!pteV(pte)) {
                pageFault(priv, vaddr, access, unsafe, "invalid entry");
                return ~0L;
            }

            if (!pteR(pte) && pteW(pte)) {
                pageFault(priv, vaddr, access, unsafe, "reserved entry type");
                return ~0L;
            }

            if (pteR(pte) || pteX(pte)) {
                pte = touch(priv, vaddr, access, unsafe, pteaddr, pte);

                final var mask = (1L << (i * 9)) - 1L;
                final var vvpn = vaddr >>> PAGE_SHIFT;

                var ppn = ppn(pte, i);
                if (i > 0) {
                    ppn &= ~mask;
                    ppn |= vvpn & mask;
                }

                final var pgsize   = 1L << (PAGE_SHIFT + i * 9);
                final var pgbase   = ppn << PAGE_SHIFT;
                final var pgoffset = vaddr & (pgsize - 1L);
                final var paddr    = pgbase | pgoffset;

                final var ptevpn = vvpn & ~mask;

                add(new Entry(pteaddr, pte, ptevpn, asid, pgbase, pgsize));

                Log.info("   => ppn=%x, pgsize=%x, pgbase=%x, pgoffset=%x, ptevpn=%x, paddr=%x",
                         ppn,
                         pgsize,
                         pgbase,
                         pgoffset,
                         ptevpn,
                         paddr);

                return paddr;
            }

            a = ppn(pte) << PAGE_SHIFT;
        }

        pageFault(priv, vaddr, access, unsafe, "missing entry");
        return ~0L;
    }

    private long sv39(
            final int priv,
            final long vaddr,
            final @NotNull Access access,
            final long asid,
            final long root,
            final boolean unsafe
    ) {
        return walk(3, priv, vaddr, access, asid, root, unsafe);
    }

    private long sv48(
            final int priv,
            final long vaddr,
            final @NotNull Access access,
            final long asid,
            final long root,
            final boolean unsafe
    ) {
        return walk(4, priv, vaddr, access, asid, root, unsafe);
    }

    private long sv57(
            final int priv,
            final long vaddr,
            final @NotNull Access access,
            final long asid,
            final long root,
            final boolean unsafe
    ) {
        return walk(5, priv, vaddr, access, asid, root, unsafe);
    }

    private void add(final @NotNull Entry entry) {
        final var asid = pteG(entry.pte) ? 0L : entry.asid;

        final var pgcount = entry.pgsize >>> PAGE_SHIFT;
        for (long i = 0L; i < pgcount; ++i) {
            final var key = new Key(entry.vpn + i, asid);
            tlb.put(key, entry);
        }
    }

    private static int getMode(final long satp) {
        return (int) ((satp >>> 60) & 0xFL);
    }

    private static int getASID(final long satp) {
        return (int) ((satp >>> 44) & 0xFFFFL);
    }

    private static long getPPN(final long satp) {
        return satp & 0xFFFFFFFFFFFL;
    }

    private static long toCause(final @NotNull Access access) {
        return switch (access) {
            case FETCH -> FETCH_PAGE_FAULT;
            case READ -> READ_PAGE_FAULT;
            case WRITE -> WRITE_PAGE_FAULT;
        };
    }

    private boolean unprivileged(final int priv, final long pte) {
        final var status = hart.csrFile().getd(sstatus, priv);
        final var sum    = (status & STATUS_SUM) != 0L;
        final var u      = pteU(pte);

        return (!sum && u && priv != CSR_U) || (!u && priv == CSR_U);
    }

    private boolean inaccessible(final int priv, final @NotNull Access access, final long pte) {

        final var u = pteU(pte);
        final var x = pteX(pte);
        final var w = pteW(pte);
        final var r = pteR(pte);

        return switch (access) {
            case FETCH -> !x || (u && priv != CSR_U);
            case READ -> !r;
            case WRITE -> !w;
        };
    }

    private void pageFault(
            final int priv,
            final long vaddr,
            final @NotNull Access access,
            final boolean unsafe,
            final @NotNull String message
    ) {
        if (unsafe) {
            Log.warn("page fault (priv=%x, vaddr=%016x, access=%s): %s",
                     priv,
                     vaddr,
                     access,
                     message);
            return;
        }
        throw new TrapException(toCause(access),
                                vaddr,
                                "page fault (priv=%x, vaddr=%016x, access=%s): %s",
                                priv,
                                vaddr,
                                access,
                                message);
    }

    private static boolean pteV(final long pte) {
        return (pte & 1L) != 0L;
    }

    private static boolean pteR(final long pte) {
        return ((pte >>> 1) & 1L) != 0L;
    }

    private static boolean pteW(final long pte) {
        return ((pte >>> 2) & 1L) != 0L;
    }

    private static boolean pteX(final long pte) {
        return ((pte >>> 3) & 1L) != 0L;
    }

    private static boolean pteU(final long pte) {
        return ((pte >>> 4) & 1L) != 0L;
    }

    private static boolean pteG(final long pte) {
        return ((pte >>> 5) & 1L) != 0L;
    }

    private static boolean pteA(final long pte) {
        return ((pte >>> 6) & 1L) != 0L;
    }

    private static boolean pteD(final long pte) {
        return ((pte >>> 7) & 1L) != 0L;
    }

    private static long vpn(final long vaddr, final int i) {
        return (vaddr >>> (PAGE_SHIFT + i * 9)) & 0x1FFL;
    }

    private static long ppn(final long pte, final int i) {
        return ((pte >>> 10) & 0xFFFFFFFFFFFL) >>> (i * 9);
    }

    private static long ppn(final long pte) {
        return (pte >>> 10) & 0xFFFFFFFFFFFL;
    }
}

File: src/main/java/io/scriptor/impl/MMU.kt
package io.scriptor.impl

import io.scriptor.isa.CSR
import io.scriptor.machine.Hart
import io.scriptor.util.Log
import io.scriptor.util.Log.format

class MMU {

    private data class Key(val vpn: ULong, val asid: ULong)

    private class Entry(
        val pteAddress: ULong,
        var pte: ULong,
        val vpn: ULong,
        val asid: ULong,
        val pgBase: ULong,
        val pgSize: ULong,
    ) {
        operator fun contains(vpn: ULong): Boolean = this.vpn <= vpn && vpn < this.vpn + (pgSize shr PAGE_SHIFT)
    }

    enum class Access {
        FETCH,
        READ,
        WRITE,
    }

    private val hart: Hart
    private val tlb: MutableMap<Key, Entry> = HashMap()

    constructor(hart: Hart) {
        this.hart = hart
    }

    fun flush(vAddress: ULong, asid: ULong) {
        if (vAddress == 0UL && asid == 0UL) {
            tlb.clear()
            return
        }

        val vpn = vAddress shr PAGE_SHIFT

        val remove = ArrayList<Key>()
        for (e in tlb) {
            val matches = (asid == 0UL || asid == e.key.asid) && (vAddress == 0UL || vpn in e.value)
            if (matches) {
                remove.add(e.key)
            }
        }

        for (key in remove) {
            tlb.remove(key)
        }
    }

    fun translate(
        privilege: UInt,
        vAddress: ULong,
        access: Access,
        unsafe: Boolean,
    ): ULong {
        if (privilege == CSR.CSR_M) {
            return vAddress
        }

        val reg = hart.csrFile[CSR.satp, privilege]
        val mode = getMode(reg)
        val asid = getASID(reg)
        val ppn = getPPN(reg)

        return when (mode) {
            0U -> vAddress
            8U -> sv39(privilege, vAddress, access, asid.toULong(), ppn, unsafe)
            9U -> sv48(privilege, vAddress, access, asid.toULong(), ppn, unsafe)
            10U -> sv57(privilege, vAddress, access, asid.toULong(), ppn, unsafe)
            else -> {
                pageFault(privilege, vAddress, access, unsafe, format("unsupported mode %d", mode))
                0UL.inv()
            }
        }
    }

    private fun touch(
        privilege: UInt,
        vAddress: ULong,
        access: Access,
        unsafe: Boolean,
        pteAddress: ULong,
        pte: ULong,
    ): ULong {
        var pte = pte
        if (inaccessible(privilege, access, pte)) {
            pageFault(privilege, vAddress, access, unsafe, "inaccessible")
        }

        if (unprivileged(privilege, pte)) {
            pageFault(privilege, vAddress, access, unsafe, "unprivileged")
        }

        if (!pteA(pte) || (access == Access.WRITE && !pteD(pte))) {
            pte = pte or (1UL shl 6)
            if (access == Access.WRITE) {
                pte = pte or (1UL shl 7)
            }
            hart.machine.pWrite(pteAddress, 8U, pte, unsafe)
        }

        return pte
    }

    private fun walk(
        levels: Int,
        privilege: UInt,
        vAddress: ULong,
        access: Access,
        asid: ULong,
        root: ULong,
        unsafe: Boolean,
    ): ULong {
        Log.info(
            "walk(levels=%d, priv=%d, vaddr=%016x, access=%s, asid=%x, root=%x)",
            levels,
            privilege,
            vAddress,
            access,
            asid,
            root,
        )

        run {
            val vpn = vAddress shr PAGE_SHIFT
            var entry = tlb[Key(vpn, asid)]
            if (entry == null) {
                entry = tlb[Key(vpn, 0UL)]
            }
            if (entry != null && vpn in entry) {
                entry.pte = touch(privilege, vAddress, access, unsafe, entry.pteAddress, entry.pte)

                val pgSize = entry.pgSize
                val pgBase = entry.pgBase
                val pgOffset = vAddress and (pgSize - 1UL)
                val pteVpn = entry.vpn
                val pAddress = pgBase or pgOffset

                Log.info(
                    "   => found translation cache: pgsize=%x, pgbase=%016x, pgoffset=%03x, ptevpn=%x, paddr=%x",
                    pgSize,
                    pgBase,
                    pgOffset,
                    pteVpn,
                    pAddress,
                )

                return pAddress
            }
        }

        var a = root shl PAGE_SHIFT

        for (i in levels - 1 downTo 0) {
            val vpn = vpn(vAddress, i)

            Log.info("  i=%d, a=%016x, vpn=%03x", i, a, vpn)

            val pteAddress = a + vpn * 8UL

            var pte = hart.machine.pRead(pteAddress, 8U, unsafe)

            Log.info(
                "   => pteaddr=%016x, pte=%016x, V=%b, R=%b, W=%b, X=%b, U=%b, G=%b, A=%b, D=%b",
                pteAddress,
                pte,
                pteV(pte),
                pteR(pte),
                pteW(pte),
                pteX(pte),
                pteU(pte),
                pteG(pte),
                pteA(pte),
                pteD(pte),
            )

            if (!pteV(pte)) {
                pageFault(privilege, vAddress, access, unsafe, "invalid entry")
                return 0UL.inv()
            }

            if (!pteR(pte) && pteW(pte)) {
                pageFault(privilege, vAddress, access, unsafe, "reserved entry type")
                return 0UL.inv()
            }

            if (pteR(pte) || pteX(pte)) {
                pte = touch(privilege, vAddress, access, unsafe, pteAddress, pte)

                val mask = (1UL shl (i * 9)) - 1UL
                val vVpn = vAddress shr PAGE_SHIFT

                var ppn = ppn(pte, i)
                if (i > 0) {
                    ppn = ppn and mask.inv()
                    ppn = ppn or (vVpn and mask)
                }

                val pgSize = 1UL shl (PAGE_SHIFT + i * 9)
                val pgBase = ppn shl PAGE_SHIFT
                val pgOffset = vAddress and (pgSize - 1UL)
                val pAddress = pgBase or pgOffset

                val pteVpn = vVpn and mask.inv()

                add(Entry(pteAddress, pte, pteVpn, asid, pgBase, pgSize))

                Log.info(
                    "   => ppn=%x, pgsize=%x, pgbase=%016x, pgoffset=%03x, ptevpn=%x, paddr=%016x",
                    ppn,
                    pgSize,
                    pgBase,
                    pgOffset,
                    pteVpn,
                    pAddress,
                )

                return pAddress
            }

            a = ppn(pte) shl PAGE_SHIFT
        }

        pageFault(privilege, vAddress, access, unsafe, "missing entry")
        return 0UL.inv()
    }

    private fun sv39(
        privilege: UInt,
        vAddress: ULong,
        access: Access,
        asid: ULong,
        root: ULong,
        unsafe: Boolean,
    ): ULong = walk(3, privilege, vAddress, access, asid, root, unsafe)

    private fun sv48(
        privilege: UInt,
        vAddress: ULong,
        access: Access,
        asid: ULong,
        root: ULong,
        unsafe: Boolean,
    ): ULong = walk(4, privilege, vAddress, access, asid, root, unsafe)

    private fun sv57(
        privilege: UInt,
        vAddress: ULong,
        access: Access,
        asid: ULong,
        root: ULong,
        unsafe: Boolean,
    ): ULong = walk(5, privilege, vAddress, access, asid, root, unsafe)

    private fun add(entry: Entry) {
        val asid = if (pteG(entry.pte)) 0UL else entry.asid

        val pgCount = entry.pgSize shr PAGE_SHIFT
        for (i in 0UL..<pgCount) {
            val key = Key(entry.vpn + i, asid)
            tlb[key] = entry
        }
    }

    private fun unprivileged(privilege: UInt, pte: ULong): Boolean {
        val status = hart.csrFile[CSR.sstatus, privilege]
        val sum = (status and CSR.STATUS_SUM) != 0UL
        val u: Boolean = pteU(pte)

        return (!sum && u && privilege != CSR.CSR_U) || (!u && privilege == CSR.CSR_U)
    }

    private fun inaccessible(privilege: UInt, access: Access, pte: ULong): Boolean {
        val u: Boolean = pteU(pte)
        val x: Boolean = pteX(pte)
        val w: Boolean = pteW(pte)
        val r: Boolean = pteR(pte)

        return when (access) {
            Access.FETCH -> !x || (u && privilege != CSR.CSR_U)
            Access.READ -> !r
            Access.WRITE -> !w
        }
    }

    private fun pageFault(
        privilege: UInt,
        vAddress: ULong,
        access: Access,
        unsafe: Boolean,
        message: String,
    ) {
        if (unsafe) {
            Log.warn(
                "page fault (priv=%d, vaddr=%016x, access=%s): %s",
                privilege,
                vAddress,
                access,
                message,
            )
            return
        }
        throw TrapException(
            hart.id,
            toCause(access),
            vAddress,
            "page fault (priv=%d, vaddr=%016x, access=%s): %s",
            privilege,
            vAddress,
            access,
            message,
        )
    }

    companion object {
        private const val FETCH_PAGE_FAULT = 0x0CUL
        private const val READ_PAGE_FAULT = 0x0DUL
        private const val WRITE_PAGE_FAULT = 0x0FUL

        private const val PAGE_SHIFT = 12

        private fun getMode(satp: ULong): UInt = ((satp shr 60) and 0xFUL).toUInt()

        private fun getASID(satp: ULong): UInt = ((satp shr 44) and 0xFFFFUL).toUInt()

        private fun getPPN(satp: ULong): ULong = satp and 0xFFFFFFFFFFFUL

        private fun toCause(access: Access): ULong = when (access) {
            Access.FETCH -> FETCH_PAGE_FAULT
            Access.READ -> READ_PAGE_FAULT
            Access.WRITE -> WRITE_PAGE_FAULT
        }

        private fun pteV(pte: ULong): Boolean = (pte and 1UL) != 0UL

        private fun pteR(pte: ULong): Boolean = ((pte shr 1) and 1UL) != 0UL

        private fun pteW(pte: ULong): Boolean = ((pte shr 2) and 1UL) != 0UL

        private fun pteX(pte: ULong): Boolean = ((pte shr 3) and 1UL) != 0UL

        private fun pteU(pte: ULong): Boolean = ((pte shr 4) and 1UL) != 0UL

        private fun pteG(pte: ULong): Boolean = ((pte shr 5) and 1UL) != 0UL

        private fun pteA(pte: ULong): Boolean = ((pte shr 6) and 1UL) != 0UL

        private fun pteD(pte: ULong): Boolean = ((pte shr 7) and 1UL) != 0UL

        private fun vpn(vAddress: ULong, i: Int): ULong = (vAddress shr (PAGE_SHIFT + i * 9)) and 0x1FFUL

        private fun ppn(pte: ULong, i: Int): ULong = ((pte shr 10) and 0xFFFFFFFFFFFUL) shr (i * 9)

        private fun ppn(pte: ULong): ULong = (pte shr 10) and 0xFFFFFFFFFFFUL
    }
}
